crypto/random/chacha20.go

```diff
@@ -0,0 +1,190 @@
+	seed [keySize]byte
+	// initial customizer
+	customizer [nonceSize]byte
+}
+
+// The main PRG, implements the Rand interface
+type chachaPRG struct {
+	genericPRG
+	core *chachaCore
+}
+
+const (
+	keySize   = chacha20.KeySize
+	nonceSize = chacha20.NonceSize
+
+	// Chacha20SeedLen is the seed length of the Chacha based PRG, it is fixed to 32 bytes.
+	Chacha20SeedLen = keySize
+	// Chacha20CustomizerMaxLen is the maximum length of the nonce used as a PRG customizer, it is fixed to 24 bytes.
+	// Shorter customizers are padded by zeros to 24 bytes.
+	Chacha20CustomizerMaxLen = nonceSize
+)
+
+// NewChacha20PRG returns a new Chacha20-based PRG, seeded with
+// the input seed (32 bytes) and a customizer (up to 12 bytes).
+//
+// It is recommended to sample the seed uniformly at random.
+// The function errors if the the seed is different than 32 bytes,
+// or if the customizer is larger than 12 bytes.
+// Shorter customizers than 12 bytes are padded by zero bytes.
+func NewChacha20PRG(seed []byte, customizer []byte) (*chachaPRG, error) {
+
+	// check the key size
+	if len(seed) != Chacha20SeedLen {
+		return nil, fmt.Errorf("chacha20 seed length should be %d, got %d", Chacha20SeedLen, len(seed))
+	}
+
+	// check the nonce size
+	if len(customizer) > Chacha20CustomizerMaxLen {
+		return nil, fmt.Errorf("new Rand streamID should be %d bytes", Chacha20CustomizerMaxLen)
+	}
+
+	// create the Chacha20 state, initialized with the seed as a key, and the customizer as a streamID.
+	chacha, err := chacha20.NewUnauthenticatedCipher(seed, customizer)
+	if err != nil {
+		return nil, fmt.Errorf("chacha20 instance creation failed: %w", err)
+	}
+
+	// init the state
+	core := &chachaCore{
+		cipher:       *chacha,
+		bytesCounter: 0,
+	}
+	copy(core.seed[:], seed)
+	copy(core.customizer[:], customizer)
+
+	prg := &chachaPRG{
+		genericPRG: genericPRG{
+			randCore: core,
+		},
+		core: core,
+	}
+	return prg, nil
+}
+
+const lenEmptyMessage = 64
+
+// Read pulls random bytes from the pseudo-random source.
+// The randoms are copied into the input buffer, the number of bytes read
+// is equal to the buffer input length.
+//
+// The stream cipher encrypts a stream of a constant message (empty for simplicity).
+func (c *chachaCore) Read(buffer []byte) {
+	// message to encrypt
+	var message []byte
+
+	if len(buffer) <= lenEmptyMessage {
+		// use a constant message (used for most of the calls)
+		message = c.emptyMessage[:len(buffer)]
+	} else {
+		// when buffer is large, use is as the message to encrypt,
+		// but this requires clearing it first.
+		for i := 0; i < len(buffer); i++ {
+			buffer[i] = 0
+		}
+		message = buffer
+	}
+	c.cipher.XORKeyStream(buffer, message)
+	// increase the counter
+	c.bytesCounter += uint64(len(buffer))
+}
+
+// Store returns the internal state of the concatenated Chacha20s
+// This is used for serialization/deserialization purposes.
+func (c *chachaPRG) Store() []byte {
+	bytes := append(c.core.seed[:], c.core.customizer[:]...)
```

Comment:
I would suggest to pre-allocate a slice of the desired length, so we avoid repeated memory allocation:
```suggestion
	bytes := make([]byte, 0, keySize+nonceSize+8)
	bytes := append(bytes, c.core.seed[:]...)
	bytes := append(bytes, c.core.customizer[:]...)
```